File: r/lab01_analysis.R
# ANLY 735 — Research Seminar in Predictive AI
# Replication Laboratory #1 | Evaluation Design Matters
# Student R Analysis
#
# Laboratory question:
# Does changing the evaluation design change estimated predictive
# performance—and does it change which model appears stronger?
#
# This script intentionally produces the analytical evidence.
# Interpretation, the FAIR audit, replication verdict, and claim boundaries
# belong in your replication-lab.qmd.
#
# Dataset:
# UCI Bike Sharing Dataset
# https://archive.ics.uci.edu/dataset/275/bike+sharing+dataset


# ============================================================
# 1. REQUIRED PACKAGES
# ============================================================

required_packages <- c(
  "readr",
  "dplyr",
  "recipes",
  "parsnip",
  "workflows",
  "yardstick",
  "ranger",
  "glmnet",
  "iterators"
)

missing_packages <- required_packages[
  !vapply(required_packages, requireNamespace, logical(1), quietly = TRUE)
]

if (length(missing_packages) > 0) {
  stop(
    paste0(
      "\nRequired R packages are missing:\n  ",
      paste(missing_packages, collapse = ", "),
      "\n\nInstall them with:\n",
      "install.packages(c(",
      paste(sprintf('"%s"', missing_packages), collapse = ", "),
      "))\n"
    ),
    call. = FALSE
  )
}

suppressPackageStartupMessages({
  library(readr)
  library(dplyr)
  library(recipes)
  library(parsnip)
  library(workflows)
  library(yardstick)
})


# ============================================================
# 2. REPRODUCIBILITY SETTINGS
# ============================================================

RANDOM_SEED <- 735
TEST_SIZE <- 0.20

set.seed(RANDOM_SEED)

# The script is designed to be run from the repository root:
#
# Rscript .\r\lab01_analysis.R
#
# or interactively from RStudio with the repository root as the
# working directory.

ROOT <- normalizePath(".", winslash = "/", mustWork = TRUE)

DATA_DIR <- file.path(ROOT, "data")
ANALYSIS_DIR <- file.path(ROOT, "analysis")

dir.create(DATA_DIR, showWarnings = FALSE, recursive = TRUE)
dir.create(ANALYSIS_DIR, showWarnings = FALSE, recursive = TRUE)

ZIP_PATH <- file.path(DATA_DIR, "Bike-Sharing-Dataset.zip")
CSV_PATH <- file.path(DATA_DIR, "hour.csv")

UCI_URL <- paste0(
  "https://archive.ics.uci.edu/ml/",
  "machine-learning-databases/00275/",
  "Bike-Sharing-Dataset.zip"
)


# ============================================================
# 3. ACQUIRE THE DATA REPRODUCIBLY
# ============================================================

if (!file.exists(CSV_PATH)) {

  cat("Downloading the UCI Bike Sharing Dataset...\n")

  tryCatch(
    {
      download.file(
        UCI_URL,
        destfile = ZIP_PATH,
        mode = "wb",
        quiet = FALSE
      )

      unzip(
        ZIP_PATH,
        files = "hour.csv",
        exdir = DATA_DIR
      )
    },
    error = function(e) {
      stop(
        paste0(
          "\nThe dataset could not be downloaded automatically.\n",
          "Check your internet connection and try again.\n",
          "Original error: ",
          conditionMessage(e)
        ),
        call. = FALSE
      )
    }
  )
}

cat("Using data:", normalizePath(CSV_PATH, winslash = "/"), "\n")


# ============================================================
# 4. LOAD DATA AND ESTABLISH CHRONOLOGY
# ============================================================

df <- read_csv(
  CSV_PATH,
  show_col_types = FALSE
)

df <- df |>
  mutate(
    datetime = as.POSIXct(
      paste(
        dteday,
        sprintf("%02d:00:00", hr)
      ),
      format = "%Y-%m-%d %H:%M:%S",
      tz = "UTC"
    )
  ) |>
  arrange(datetime)

cat("\nDataset dimensions:\n")
cat(nrow(df), "rows x", ncol(df), "columns\n")

cat("\nObservation period:\n")
cat(
  format(min(df$datetime), "%Y-%m-%d %H:%M:%S"),
  "to",
  format(max(df$datetime), "%Y-%m-%d %H:%M:%S"),
  "\n"
)


# ============================================================
# 5. DEFINE THE PREDICTION PROBLEM
# ============================================================

TARGET <- "cnt"

CATEGORICAL_FEATURES <- c(
  "season",
  "yr",
  "mnth",
  "hr",
  "holiday",
  "weekday",
  "workingday",
  "weathersit"
)

NUMERIC_FEATURES <- c(
  "temp",
  "atemp",
  "hum",
  "windspeed"
)

FEATURES <- c(
  CATEGORICAL_FEATURES,
  NUMERIC_FEATURES
)

# IMPORTANT:
# "casual" and "registered" are deliberately excluded.
# Their sum defines the target variable "cnt".
#
# "instant" is an identifier.
# "dteday" establishes chronology but is not used directly as a predictor.


# ============================================================
# 6. PREPARE MODELING DATA
# ============================================================

model_df <- df |>
  select(
    datetime,
    all_of(FEATURES),
    all_of(TARGET)
  )

# Convert categorical predictors to factors so that the R analysis
# treats them as categorical variables, matching the Python track's
# one-hot encoding strategy.

model_df <- model_df |>
  mutate(
    across(
      all_of(CATEGORICAL_FEATURES),
      as.factor
    )
  )


# ============================================================
# 7. DEFINE PREPROCESSING
# ============================================================

make_recipe <- function(training_data) {

  recipe(
    cnt ~ .,
    data = training_data |>
      select(-datetime)
  ) |>
    step_impute_median(all_numeric_predictors()) |>
    step_impute_mode(all_nominal_predictors()) |>
    step_dummy(all_nominal_predictors(), one_hot = TRUE) |>
    step_normalize(all_numeric_predictors())
}


# ============================================================
# 8. DEFINE THE TWO COMPETING MODELS
# ============================================================

ridge_spec <- linear_reg(
  penalty = 1.0,
  mixture = 0
) |>
  set_engine("glmnet") |>
  set_mode("regression")

rf_spec <- rand_forest(
  trees = 300,
  min_n = 2
) |>
  set_engine(
    "ranger",
    seed = RANDOM_SEED,
    num.threads = max(1, parallel::detectCores() - 1)
  ) |>
  set_mode("regression")


# ============================================================
# 9. EVALUATION FUNCTION
# ============================================================

evaluate_model <- function(
  model_spec,
  train_data,
  test_data,
  evaluation_design,
  model_name
) {

  rec <- make_recipe(train_data)

  wf <- workflow() |>
    add_recipe(rec) |>
    add_model(model_spec)

  fitted_model <- fit(
    wf,
    data = train_data |> select(-datetime)
  )

  predictions <- predict(
    fitted_model,
    new_data = test_data |> select(-datetime)
  )

  evaluated <- bind_cols(
    test_data |> select(cnt),
    predictions
  )

  rmse_value <- rmse(
    evaluated,
    truth = cnt,
    estimate = .pred
  ) |>
    pull(.estimate)

  mae_value <- mae(
    evaluated,
    truth = cnt,
    estimate = .pred
  ) |>
    pull(.estimate)

  tibble(
    `Evaluation Design` = evaluation_design,
    Model = model_name,
    RMSE = rmse_value,
    MAE = mae_value
  )
}


# ============================================================
# 10. EXPERIMENT 1 — RANDOM EVALUATION
# ============================================================

set.seed(RANDOM_SEED)

n <- nrow(model_df)
test_n <- ceiling(n * TEST_SIZE)

test_indices <- sample(
  seq_len(n),
  size = test_n,
  replace = FALSE
)

train_random <- model_df[-test_indices, ]
test_random <- model_df[test_indices, ]

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("EXPERIMENT 1 — RANDOM EVALUATION\n")
cat(paste(rep("=", 60), collapse = ""), "\n", sep = "")

cat("Training observations:", nrow(train_random), "\n")
cat("Testing observations :", nrow(test_random), "\n")

random_ridge <- evaluate_model(
  ridge_spec,
  train_random,
  test_random,
  "Random",
  "Ridge Regression"
)

random_rf <- evaluate_model(
  rf_spec,
  train_random,
  test_random,
  "Random",
  "Random Forest"
)


# ============================================================
# 11. EXPERIMENT 2 — TEMPORAL EVALUATION
# ============================================================

# Match the Python design:
# floor(80% of observations) are used for training and the
# remaining observations form the future test period.

split_index <- floor(n * (1 - TEST_SIZE))

train_temporal <- model_df[seq_len(split_index), ]
test_temporal <- model_df[(split_index + 1):n, ]

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("EXPERIMENT 2 — TEMPORAL EVALUATION\n")
cat(paste(rep("=", 60), collapse = ""), "\n", sep = "")

cat("\nTraining period:\n")
cat(
  format(min(train_temporal$datetime), "%Y-%m-%d %H:%M:%S"),
  "to",
  format(max(train_temporal$datetime), "%Y-%m-%d %H:%M:%S"),
  "\n"
)

cat("\nTesting period:\n")
cat(
  format(min(test_temporal$datetime), "%Y-%m-%d %H:%M:%S"),
  "to",
  format(max(test_temporal$datetime), "%Y-%m-%d %H:%M:%S"),
  "\n"
)

cat("\nTraining observations:", nrow(train_temporal), "\n")
cat("Testing observations :", nrow(test_temporal), "\n")

temporal_ridge <- evaluate_model(
  ridge_spec,
  train_temporal,
  test_temporal,
  "Temporal",
  "Ridge Regression"
)

temporal_rf <- evaluate_model(
  rf_spec,
  train_temporal,
  test_temporal,
  "Temporal",
  "Random Forest"
)


# ============================================================
# 12. DISPLAY THE EVIDENCE
# ============================================================

results_df <- bind_rows(
  random_ridge,
  random_rf,
  temporal_ridge,
  temporal_rf
) |>
  mutate(
    RMSE = round(RMSE, 2),
    MAE = round(MAE, 2)
  )

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("ANLY 735 — EVALUATION RESULTS\n")
cat(paste(rep("=", 60), collapse = ""), "\n\n", sep = "")

print(results_df)


# ============================================================
# 13. DISPLAY MODEL RANKINGS
# ============================================================

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("MODEL RANKING BY RMSE\n")
cat(paste(rep("=", 60), collapse = ""), "\n", sep = "")

for (design in c("Random", "Temporal")) {

  ranked <- results_df |>
    filter(`Evaluation Design` == design) |>
    arrange(RMSE)

  cat("\n", design, " evaluation:\n", sep = "")

  for (i in seq_len(nrow(ranked))) {
    cat(
      i,
      ". ",
      ranked$Model[i],
      " (RMSE = ",
      ranked$RMSE[i],
      ", MAE = ",
      ranked$MAE[i],
      ")\n",
      sep = ""
    )
  }
}


# ============================================================
# 14. SAVE REPRODUCIBLE RESULTS
# ============================================================

output_path <- file.path(
  ANALYSIS_DIR,
  "lab01_results.csv"
)

write_csv(
  results_df,
  output_path
)

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("RESULTS SAVED\n")
cat(paste(rep("=", 60), collapse = ""), "\n", sep = "")

cat(normalizePath(output_path, winslash = "/"), "\n")


# ============================================================
# 15. RESEARCHER HANDOFF
# ============================================================

cat("\n", paste(rep("=", 60), collapse = ""), "\n", sep = "")
cat("YOUR JOB AS THE RESEARCHER\n")
cat(paste(rep("=", 60), collapse = ""), "\n", sep = "")

cat(
  "
The script has produced evidence. It has not interpreted that evidence for you.

Return to replication-lab.qmd and investigate:

1. How did RMSE and MAE change across evaluation designs?
2. Did the model ranking change?
3. Which conclusions remained stable?
4. Which conclusions require qualification?
5. Which evaluation design better represents the intended deployment scenario?
6. What does the evidence permit you to claim?
7. What does the evidence NOT permit you to claim?

Do not chase identical numbers. Investigate reproducibility.
"
)

File: python/lab01_analysis.py
"""
ANLY 735 — Research Seminar in Predictive AI
Replication Laboratory #1 | Evaluation Design Matters
Student Python Analysis

Laboratory question:
Does changing the evaluation design change estimated predictive
performance—and does it change which model appears stronger?

This script intentionally produces the analytical evidence.
Interpretation, the FAIR audit, replication verdict, and claim boundaries
belong in your replication-lab.qmd.

Dataset:
UCI Bike Sharing Dataset
https://archive.ics.uci.edu/dataset/275/bike+sharing+dataset
"""

from pathlib import Path
import urllib.request
import zipfile

import numpy as np
import pandas as pd

from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestRegressor
from sklearn.impute import SimpleImputer
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_absolute_error, mean_squared_error
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import OneHotEncoder, StandardScaler


# ============================================================
# 1. REPRODUCIBILITY SETTINGS
# ============================================================

RANDOM_SEED = 735
TEST_SIZE = 0.20

np.random.seed(RANDOM_SEED)

ROOT = Path(__file__).resolve().parents[1]
DATA_DIR = ROOT / "data"
ANALYSIS_DIR = ROOT / "analysis"

DATA_DIR.mkdir(exist_ok=True)
ANALYSIS_DIR.mkdir(exist_ok=True)

ZIP_PATH = DATA_DIR / "Bike-Sharing-Dataset.zip"
CSV_PATH = DATA_DIR / "hour.csv"

UCI_URL = (
    "https://archive.ics.uci.edu/ml/"
    "machine-learning-databases/00275/"
    "Bike-Sharing-Dataset.zip"
)


# ============================================================
# 2. ACQUIRE THE DATA REPRODUCIBLY
# ============================================================

if not CSV_PATH.exists():
    print("Downloading the UCI Bike Sharing Dataset...")

    try:
        urllib.request.urlretrieve(UCI_URL, ZIP_PATH)

        with zipfile.ZipFile(ZIP_PATH, "r") as archive:
            archive.extract("hour.csv", DATA_DIR)

    except Exception as exc:
        raise RuntimeError(
            "\nThe dataset could not be downloaded automatically.\n"
            "Check your internet connection and try again.\n"
            f"Original error: {exc}"
        ) from exc

print(f"Using data: {CSV_PATH}")


# ============================================================
# 3. LOAD DATA AND ESTABLISH CHRONOLOGY
# ============================================================

df = pd.read_csv(CSV_PATH)

df["datetime"] = pd.to_datetime(
    df["dteday"].astype(str)
    + " "
    + df["hr"].astype(str).str.zfill(2)
    + ":00:00"
)

df = df.sort_values("datetime").reset_index(drop=True)

print("\nDataset dimensions:")
print(df.shape)

print("\nObservation period:")
print(f"{df['datetime'].min()} to {df['datetime'].max()}")


# ============================================================
# 4. DEFINE THE PREDICTION PROBLEM
# ============================================================

TARGET = "cnt"

CATEGORICAL_FEATURES = [
    "season",
    "yr",
    "mnth",
    "hr",
    "holiday",
    "weekday",
    "workingday",
    "weathersit",
]

NUMERIC_FEATURES = [
    "temp",
    "atemp",
    "hum",
    "windspeed",
]

FEATURES = CATEGORICAL_FEATURES + NUMERIC_FEATURES

# IMPORTANT:
# "casual" and "registered" are deliberately excluded.
# Their sum defines the target variable "cnt".
#
# "instant" is an identifier.
# "dteday" establishes chronology but is not used directly as a predictor.

X = df[FEATURES]
y = df[TARGET]


# ============================================================
# 5. DEFINE PREPROCESSING
# ============================================================

def make_preprocessor():
    """
    Create a fresh preprocessing pipeline.

    Numeric predictors:
      - median imputation
      - standardization

    Categorical predictors:
      - most-frequent imputation
      - one-hot encoding

    Preprocessing is fitted only on the training data because it is
    contained inside each scikit-learn Pipeline.
    """

    numeric_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="median")),
            ("scaler", StandardScaler()),
        ]
    )

    categorical_transformer = Pipeline(
        steps=[
            ("imputer", SimpleImputer(strategy="most_frequent")),
            (
                "onehot",
                OneHotEncoder(handle_unknown="ignore"),
            ),
        ]
    )

    return ColumnTransformer(
        transformers=[
            ("num", numeric_transformer, NUMERIC_FEATURES),
            ("cat", categorical_transformer, CATEGORICAL_FEATURES),
        ]
    )


# ============================================================
# 6. DEFINE THE TWO COMPETING MODELS
# ============================================================

def make_ridge():
    """Return a fresh Ridge Regression pipeline."""

    return Pipeline(
        steps=[
            ("preprocess", make_preprocessor()),
            ("model", Ridge(alpha=1.0)),
        ]
    )


def make_random_forest():
    """Return a fresh Random Forest pipeline."""

    return Pipeline(
        steps=[
            ("preprocess", make_preprocessor()),
            (
                "model",
                RandomForestRegressor(
                    n_estimators=300,
                    min_samples_leaf=2,
                    random_state=RANDOM_SEED,
                    n_jobs=-1,
                ),
            ),
        ]
    )


# ============================================================
# 7. EVALUATION FUNCTION
# ============================================================

def evaluate_model(
    model,
    X_train,
    y_train,
    X_test,
    y_test,
    evaluation_design,
    model_name,
):
    """
    Fit a model on training data and evaluate it on untouched test data.

    Returns RMSE and MAE. Lower values indicate smaller prediction errors.
    """

    model.fit(X_train, y_train)

    predictions = model.predict(X_test)

    rmse = np.sqrt(mean_squared_error(y_test, predictions))
    mae = mean_absolute_error(y_test, predictions)

    return {
        "Evaluation Design": evaluation_design,
        "Model": model_name,
        "RMSE": rmse,
        "MAE": mae,
    }


results = []


# ============================================================
# 8. EXPERIMENT 1 — RANDOM EVALUATION
# ============================================================

X_train_random, X_test_random, y_train_random, y_test_random = (
    train_test_split(
        X,
        y,
        test_size=TEST_SIZE,
        random_state=RANDOM_SEED,
    )
)

print("\n" + "=" * 60)
print("EXPERIMENT 1 — RANDOM EVALUATION")
print("=" * 60)

print(f"Training observations: {len(X_train_random)}")
print(f"Testing observations : {len(X_test_random)}")

results.append(
    evaluate_model(
        make_ridge(),
        X_train_random,
        y_train_random,
        X_test_random,
        y_test_random,
        "Random",
        "Ridge Regression",
    )
)

results.append(
    evaluate_model(
        make_random_forest(),
        X_train_random,
        y_train_random,
        X_test_random,
        y_test_random,
        "Random",
        "Random Forest",
    )
)


# ============================================================
# 9. EXPERIMENT 2 — TEMPORAL EVALUATION
# ============================================================

split_index = int(len(df) * (1 - TEST_SIZE))

train_temporal = df.iloc[:split_index]
test_temporal = df.iloc[split_index:]

X_train_temporal = train_temporal[FEATURES]
y_train_temporal = train_temporal[TARGET]

X_test_temporal = test_temporal[FEATURES]
y_test_temporal = test_temporal[TARGET]

print("\n" + "=" * 60)
print("EXPERIMENT 2 — TEMPORAL EVALUATION")
print("=" * 60)

print("\nTraining period:")
print(
    f"{train_temporal['datetime'].min()} "
    f"to {train_temporal['datetime'].max()}"
)

print("\nTesting period:")
print(
    f"{test_temporal['datetime'].min()} "
    f"to {test_temporal['datetime'].max()}"
)

print(f"\nTraining observations: {len(train_temporal)}")
print(f"Testing observations : {len(test_temporal)}")

results.append(
    evaluate_model(
        make_ridge(),
        X_train_temporal,
        y_train_temporal,
        X_test_temporal,
        y_test_temporal,
        "Temporal",
        "Ridge Regression",
    )
)

results.append(
    evaluate_model(
        make_random_forest(),
        X_train_temporal,
        y_train_temporal,
        X_test_temporal,
        y_test_temporal,
        "Temporal",
        "Random Forest",
    )
)


# ============================================================
# 10. DISPLAY THE EVIDENCE
# ============================================================

results_df = pd.DataFrame(results)

display_results = results_df.copy()
display_results["RMSE"] = display_results["RMSE"].round(2)
display_results["MAE"] = display_results["MAE"].round(2)

print("\n" + "=" * 60)
print("ANLY 735 — EVALUATION RESULTS")
print("=" * 60 + "\n")

print(display_results.to_string(index=False))


# ============================================================
# 11. DISPLAY MODEL RANKINGS
# ============================================================

print("\n" + "=" * 60)
print("MODEL RANKING BY RMSE")
print("=" * 60)

for design in ["Random", "Temporal"]:

    subset = (
        display_results[
            display_results["Evaluation Design"] == design
        ]
        .sort_values("RMSE")
        .reset_index(drop=True)
    )

    print(f"\n{design} evaluation:")

    for position, row in subset.iterrows():
        print(
            f"{position + 1}. {row['Model']} "
            f"(RMSE = {row['RMSE']}, MAE = {row['MAE']})"
        )


# ============================================================
# 12. SAVE REPRODUCIBLE RESULTS
# ============================================================

output_path = ANALYSIS_DIR / "lab01_results.csv"

display_results.to_csv(output_path, index=False)

print("\n" + "=" * 60)
print("RESULTS SAVED")
print("=" * 60)

print(output_path)


# ============================================================
# 13. RESEARCHER HANDOFF
# ============================================================

print("\n" + "=" * 60)
print("YOUR JOB AS THE RESEARCHER")
print("=" * 60)

print(
    """
The script has produced evidence. It has not interpreted that evidence for you.

Return to replication-lab.qmd and investigate:

1. How did RMSE and MAE change across evaluation designs?
2. Did the model ranking change?
3. Which conclusions remained stable?
4. Which conclusions require qualification?
5. Which evaluation design better represents the intended deployment scenario?
6. What does the evidence permit you to claim?
7. What does the evidence NOT permit you to claim?

Do not chase identical numbers. Investigate reproducibility.
"""
)
